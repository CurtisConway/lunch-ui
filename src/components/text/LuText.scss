@import '../../assets/sass/_variables';
@import '../../assets/sass/_mixins';

.lu-text {
  padding:0;
  margin:0;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  text-rendering: optimizeLegibility;

  @media
  (-webkit-min-device-pixel-ratio: 2),
  (min-resolution: 192dpi) {
    -webkit-font-smoothing:subpixel-antialiased;
  }
}

.text-uppercase {
  text-transform:uppercase;
}

.text-capitalize {
  text-transform:capitalize;
}