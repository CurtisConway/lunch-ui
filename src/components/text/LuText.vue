<template>
  <component
    :is="tag"
    :class="defaultClassList"
    :style="defaultStyleList"
  >
    <slot></slot>
  </component>
</template>

<script>
import './LuText.scss';
import colors from '../../assets/js/mixins/colors';
import spacing from '../../assets/js/mixins/spacing';

export default {
  name: 'LuText',
  mixins: [colors, spacing],
  props: {
    tag: {
      type: String,
      default: () => 'p',
    },
    transform: {
      type: String,
      default: () => undefined,
    },
    size: {
      type: [String, Number],
      default: () => 'body',
    },
    color: {
      type: String,
      default: () => 'black',
    },
    bold: {
      type: Boolean,
      default: () => false,
    },
    italic: {
      type: Boolean,
      default: () => false,
    },
    underline: {
      type: Boolean,
      default: () => false,
    },
    align: {
      type: String,
      default: () => 'left',
    },
  },
  computed: {
    defaultClassList() {
      return [
        'lu-text',
        this.textColor,
        this.textTransformClass,
        this.marginClasses,
        this.paddingClasses,
        this.textSizeClass,
        this.textAlignClass,
        {
          'text-bold': this.bold,
          'text-italic': this.italic,
          'text-underline': this.underline,
        },
      ];
    },
    defaultStyleList() {
      return [
        this.textSizeStyle,
      ];
    },
    textTransformClass() {
      const transforms = {
        uppercase: 'text-uppercase',
        capitalize: 'text-capitalize',
      };
      return transforms[this.transform] || '';
    },
    textSizeClass() {
      if (typeof this.size === 'number') {
        return false;
      }
      return `text-${this.size}`;
    },
    textSizeStyle() {
      if (typeof this.size !== 'number') {
        return false;
      }
      return { 'font-size': `${this.size}px` };
    },
    textAlignClass() {
      return {
        left: 'text-left',
        center: 'text-center',
        right: 'text-right',
      }[this.align];
    },
  },
};
</script>
