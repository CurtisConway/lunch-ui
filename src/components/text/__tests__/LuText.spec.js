import Vue from 'vue';
import Router from 'vue-router';

import LuText from '../LuText.vue';

import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import { compileToFunctions } from 'vue-template-compiler'

describe('LuText.vue', () => {
  let mountFunction;
  let router = Router;
  let localVue = typeof Vue;

  beforeEach(() => {
    router = new Router();
    localVue = createLocalVue();
    localVue.use(Router);

    mountFunction = (options = {}) => {
      return mount(LuText, {
        localVue,
        router,
        ...options,
      });
    }
  })

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot();
  });

  it('should render component with a tag prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        tag: 'h2',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with a color prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        color: 'green',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with a transform prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        transform: 'uppercase',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with a size prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        size: 'heading',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with a bold prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        bold: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with an underline prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        underline: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should render component with an italic prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        italic: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should have the correct text color class', () => {
    const wrapper = mountFunction({
      propsData: {
        color: 'green',
      },
    });

    expect(wrapper.classes('text-green')).toBe(true);
  });

  it('should have the correct text transform class', () => {
    const wrapper = mountFunction({
      propsData: {
        transform: 'uppercase',
      },
    });

    expect(wrapper.classes('text-uppercase')).toBe(true);
  });

  it('should have the correct font size class', () => {
    const wrapper = mountFunction({
      propsData: {
        size: 'heading',
      },
    });

    expect(wrapper.classes('text-heading')).toBe(true);
  });

  it('should have the proper styles if size is a number', async () => {
    const wrapper = mountFunction({
      propsData: {
        size: 20,
      },
    });

    expect(wrapper.attributes().style)
      .toBe('font-size: 20px;');
  });

  it('should have the text-bold class if bold is set', () => {
    const wrapper = mountFunction({
      propsData: {
        bold: true,
      },
    });

    expect(wrapper.classes('text-bold')).toBe(true);
  });

  it('should have the text-italic class if italic is set', () => {
    const wrapper = mountFunction({
      propsData: {
        italic: true,
      },
    });

    expect(wrapper.classes('text-italic')).toBe(true);
  });

  it('should have the text-underline class if underline is set', () => {
    const wrapper = mountFunction({
      propsData: {
        underline: true,
      },
    });

    expect(wrapper.classes('text-underline')).toBe(true);
  });
});
