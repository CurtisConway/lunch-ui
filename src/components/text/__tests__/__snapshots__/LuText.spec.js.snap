// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuText.vue should render component and match snapshot 1`] = `<p class="lu-text text-black text-body"></p>`;

exports[`LuText.vue should render component with a bold prop and match snapshot 1`] = `<p class="lu-text text-black text-body text-bold"></p>`;

exports[`LuText.vue should render component with a color prop and match snapshot 1`] = `<p class="lu-text text-green text-body"></p>`;

exports[`LuText.vue should render component with a size prop and match snapshot 1`] = `<p class="lu-text text-black text-heading"></p>`;

exports[`LuText.vue should render component with a tag prop and match snapshot 1`] = `<h2 class="lu-text text-black text-body"></h2>`;

exports[`LuText.vue should render component with a transform prop and match snapshot 1`] = `<p class="lu-text text-black text-uppercase text-body"></p>`;

exports[`LuText.vue should render component with an italic prop and match snapshot 1`] = `<p class="lu-text text-black text-body text-italic"></p>`;

exports[`LuText.vue should render component with an underline prop and match snapshot 1`] = `<p class="lu-text text-black text-body text-underline"></p>`;
